runtime-modules/content-directory/src/lib.rs

```diff
@@ -0,0 +1,1645 @@
+        /// class limit default value.
+        pub fn create_entity(
+            origin,
+            class_id: T::ClassId,
+            actor: Actor<T>,
+        ) -> dispatch::Result {
+            let account_id = ensure_signed(origin)?;
+            let class = Self::ensure_class_exists(class_id)?;
+            class.ensure_maximum_entities_count_limit_not_reached()?;
+
+            let class_permissions = class.get_permissions();
+            class_permissions.ensure_entity_creation_not_blocked()?;
+            class_permissions.ensure_can_create_entities(&account_id, &actor)?;
+
+            let entity_controller = EntityController::from_actor(&actor);
+
+            // Check if entity creation voucher exists
+            let voucher_exists = if <EntityCreationVouchers<T>>::exists(class_id, &entity_controller) {
+                let entity_creation_voucher = Self::entity_creation_vouchers(class_id, &entity_controller);
+
+                // Ensure voucher limit not reached
+                Self::ensure_voucher_limit_not_reached(entity_creation_voucher)?;
+                true
+            } else {
+                false
+            };
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            if voucher_exists {
+                // Increment number of created entities count, if voucher already exist
+                <EntityCreationVouchers<T>>::mutate(class_id, &entity_controller, |entity_creation_voucher| {
+                    entity_creation_voucher.increment_created_entities_count()
+                });
+            } else {
+                // Create new voucher for given entity creator with default limit and increment created entities count
+                let mut entity_creation_voucher = EntityCreationVoucher::new(class.get_controller_entity_creation_limit());
+                entity_creation_voucher.increment_created_entities_count();
+                <EntityCreationVouchers<T>>::insert(class_id, entity_controller.clone(), entity_creation_voucher);
+            }
+
+            Self::complete_entity_creation(class_id, entity_controller);
+            Ok(())
+        }
+
+        pub fn remove_entity(
+            origin,
+            actor: Actor<T>,
+            entity_id: T::EntityId,
+        ) -> dispatch::Result {
+            let account_id = ensure_signed(origin)?;
+
+            Self::ensure_known_entity_id(entity_id)?;
+
+            let (_, access_level) = Self::get_entity_and_access_level(account_id, entity_id, actor)?;
+            EntityPermissions::<T>::ensure_group_can_remove_entity(access_level)?;
+
+            // Ensure there is no property values pointing to the given entity
+            Self::ensure_rc_is_zero(entity_id)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            Self::complete_entity_removal(entity_id);
+
+            Ok(())
+        }
+
+        pub fn add_schema_support_to_entity(
```

Comment:
I am unable to convince myself that this is correct, despite looking at it now for 25 mins, so this has to be simplified.

Reply:
Good point, did not refactored this extrinsic much https://github.com/Joystream/joystream/pull/470/commits/a59e68670b1d533f130014a5baf1c8d7ee923f33